### PR TITLE
decoder only transformer

### PR DESCRIPTION
**Pull request by hurryman42**
addresses #27 and also #26 but needs review
training started successfully, but I didn't let it finish (would have taken too long), will try to get GPU support, otherwise testing on smaller dataset is an option for now

### CODE REVIEW

File: src/models/transformer/feed_forward.py [hunk 1]
```diff
@@ -2,9 +2,14 @@
 
 
 class FeedForward(nn.Module):
-    def __init__(self):
+    def __init__(self, embedding_dimension, ff_hidden_dimension, dropout=0.1):
```
Comment (TillProjects) [inline src/models/transformer/feed_forward.py]: vielleicht mit `ff_hidden_dim=None` und hinter den `super().__init__()` `ff_hidden_dimension = ff_hidden_dimension or 4 * embedding_dimension` 

bzw. ich würde das im Block oder noch weiter oben anpassen

4 mal embedding_dimension ist wohl so der typische default und dann können wir bei Bedarf anpassen

File: src/models/transformer/attention.py [hunk 2]
```diff
@@ -1,19 +1,42 @@
...
 
 class Head(nn.Module):
-    def __init__(self):
+    def __init__(self, embedding_dimension, head_dimension, block_size):
         super().__init__()
-        pass
+        self.key = nn.Linear(embedding_dimension, head_dimension, bias=False)
+        self.query = nn.Linear(embedding_dimension, head_dimension, bias=False)
+        self.value = nn.Linear(embedding_dimension, head_dimension, bias=False)
+        #self.register_buffer("triu", torch.triu(torch.ones(block_size, block_size)))
```
Comment (TillProjects) [inline src/models/transformer/attention.py]: `register_buffer` ist besser, weil sonst die Maske jedes mal neuberechnet wird

im forward kannst du dann mit `attention_scores = attention_scores.masked_fill(self.mask[:seq_length, :seq_length], float('-inf'))` die zu recht schneiden

File: src/scripts/train_transformer.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    print(f"Using device: {device}\n")
+
+    # prepare data
+    encoded_texts = encode_text(text, char_tokenizer)
+    train_data, val_data = train_val_split(encoded_texts, 0.9)
+
+    # initialize model
+    model = TransformerDecoderOnly(
+        vocab_size, embedding_dimension=DIMENSION_MODEL, num_blocks=6, num_heads=8,
+        head_dimension=32, block_size=SEQ_LEN, ff_hidden_dimension=1024, dropout=0.1
```
Comment (TillProjects) [inline src/scripts/train_transformer.py]: head_dimension = embedding_dimension // num_heads

File: src/models/transformer/attention.py [hunk 4]
```diff
@@ -1,19 +1,42 @@
...
 
 class MultiHeadAttention(nn.Module):
-    def __init__(self):
+    def __init__(self, num_heads, embedding_dimension, head_dimension, block_size):
         super().__init__()
-        pass
+        self.heads = nn.ModuleList([
+            Head(embedding_dimension, head_dimension, block_size) for _ in range(num_heads)
+        ])
+        self.proj = nn.Linear(num_heads * head_dimension, embedding_dimension) # output projection: combine multiple heads
```
Comment (TillProjects) [inline src/models/transformer/attention.py]: `nn.Linear(num_heads * head_dimension, embedding_dimension)`
wenn du die Aenderung mit `embedding_dimension // num_heads` machtst

File: src/scripts/train_char_tokenizer.py [hunk 5]
```diff
@@ -1,13 +1,13 @@
...
-from src.scripts.read_file import read_file
+from src.scripts.read_file import read_file_only_reviews, read_file_synopsis_review_pairs
 from src.tokenizer.char_tokenizer import CharTokenizer
 
 if __name__ == "__main__":
     BASE_DIR = Path(__file__).resolve().parent.parent
     input_file = BASE_DIR.parent / "data" / "letterboxd_filtered_short_synopsis_film.jsonl"
 
-    texts = read_file(input_file)
+    texts = read_file_synopsis_review_pairs(input_file)
```
Comment (TillProjects) [inline src/scripts/train_char_tokenizer.py]: `read_file_only_reviews` auch hier erstmal nur level 1

File: src/models/transformer/transformer.py [hunk 6]
```diff
@@ -1,13 +1,53 @@
...
 
-    def generate(self):
-        pass
+    def forward(self, index):
+        batch_size, seq_length = index.shape
+        assert seq_length <= self.block_size, f"Cannot forward sequence of length {seq_length}, block size is only {self.block_size}"
+
+        token_emb = self.token_embedding(index)                 # (batch_size, seq_length, embed_dim)
+        pos = torch.arange(seq_length, device=index.device)     # (seq_length,)
+        pos_emb = self.position_embedding(pos)[None, :, :]      # (1, seq_length, embedding_dimension)
```
Comment (TillProjects) [inline src/models/transformer/transformer.py]: alternativ glaube ich: `pos_emb = self.position_embedding(pos).unsqueeze(0)`

File: src/scripts/train_transformer.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    eval_iters: int,
+    device: str,
+) -> dict:
+    out = {}
+    model.eval()
+    for split in ["train", "val"]:
+        losses = torch.zeros(eval_iters)
+        for k in range(eval_iters):
+            x, y = get_batch(split, train_data, val_data, seq_len, batch_size, device)
+            print(x.min(), x.max())
```
[the remark this answers is not in the document]
Comment (hurryman42) [reply src/scripts/train_transformer.py]: meinst du das print oder noch mehr?

File: src/models/transformer/attention.py [hunk 8]
```diff
@@ -1,19 +1,42 @@
+import torch
 import torch.nn as nn
+import torch.nn.functional as F
 
 
 class Head(nn.Module):
```
Comment (TillProjects) [inline src/models/transformer/attention.py]: vielleich das file oder den head in SelfAttentionHead umbenenne, da wir hier Self-Attention benutzten

File: src/models/transformer/attention.py (hunk 4 above)
Comment (TillProjects) [reply src/models/transformer/attention.py]: ups meinte `embedding_dimension, embedding_dimension` als Argument
